osf/metrics/reporters/spam_count.py

```diff
@@ -0,0 +1,84 @@
+from osf.models import OSFUser
+
+from osf.metrics.reports import SpamSummaryReport
+from ._base import MonthlyReporter
+from osf.models import PreprintLog, NodeLog
+from osf.models.spam import SpamStatus
+
+
+class SpamCountReporter(MonthlyReporter):
+
+    def report(self, report_yearmonth):
+        target_month = report_yearmonth.target_month()
+        next_month = report_yearmonth.next_month()
+
```

Comment:
Future improvement?
Could probably run all of the nodelog aggregations here and then just filter on that queryset below instead of issuing a bunch of nearly-identical queries.

Reply:
yeah should be able to [use `aggregate`](https://docs.djangoproject.com/en/3.2/ref/models/conditional-expressions/#conditional-aggregation) to run only one query per table, e.g.
```py
node_log_qs = NodeLog.objects.filter(created__range=(target_month, next_month))
node_log_counts = node_log_qs.aggregate(
    nodes_confirmed_spam=Count('pk', filter=Q(action=NodeLog.CONFIRM_SPAM, node__type='osf.node')),
    nodes_confirmed_ham=Count('pk', filter=Q(action=NodeLog.CONFIRM_HAM, node__type='osf.node')),
    # ...
)
```

Reply:
Oh, yeah, and some things that that have been `confirmed_ham` may have previously been `confirmed_spam` depending on what our spam settings where at the time (flag vs auto-spam) -- and Flagged will likely include lots of things in both the confirmed_ham and confirmed_spam sets. That may be ok, but depends on what Product is expecting. Seems like something that's worth a quick sync unless expectations are explicitly documented somewhere. In any case, probably worth trying something like:

```
spam_nodelogs = NodeLog.objects.filter(
  created__gt=target_month,
  created__lt=next_month,
  Q(action=NodeLog.CONFIRM_SPAM) | Q(action=NodeLog.CONFIRM_HAM) | Q(action=NodeLog.FLAG_SPAM)
)
spammed_nodes =  AbstractNode.objects.annotate(
  confirmed_spam=Exists(spam_nodelogs.filter(node__id=OuterRef('id')),
  confirmed_ham=Exists(spam_nodelogs.filter(node__id=OuterRef('id')),
  flagged_spam=Exists(spam_nodelogs.filter(node__id=OuterRef('id')),
).aggregate(
  nodes_confirmed_spam=Count('pk', filter=Q(confirmd_spam=True, confirmed_ham=False, type='osf.node'),
  nodes_confirmed_ham=Count('pk', filter=Q(confirmed_ham=True, type='osf.node')
  registrations_confirmed_spam=#. . .
)
```

Reply:
I have gone over this with Product, they expect the account given by the nodelogs and I explained the difference in the OSFUser spam data. It would be nice to know both of these things though.